skills/dff_reasoning_skill/scenario/response.py

```diff
@@ -217,79 +256,158 @@ def response_with_chosen_api(ctx: Context, actor: Actor, *args, **kwargs) -> str
+        int_ctx.save_to_shared_memory(ctx, actor, api2use=api2use)
+        logger.info(f"CHOSEN TOOL: {api2use}")
+        time.sleep(5)
+        return api2use
+
+
+def ask4approval(ctx: Context, actor: Actor, *args, **kwargs) -> str:
+    if not ctx.validation:
+        shared_memory = int_ctx.get_shared_memory(ctx, actor)
+        api2use = shared_memory.get("api2use", None)
+        response = f"""I need to use {api_conf[api2use]['display_name']} \
 to handle your request. Do you approve?"""
-                except KeyError:
-                    for key in api_conf.keys():
-                        if key in hypotheses[0]:
-                            if api_conf[key]["needs_approval"] == "False":
-                                api2use = key
-                                int_ctx.save_to_shared_memory(ctx, actor, api2use=api2use)
-                                timeout = api_conf[api2use]["timeout"]
-                                signal.signal(signal.SIGALRM, timeout_handler)
-                                signal.alarm(timeout)
-                                try:
-                                    response = globals()[f"{api2use}_response"](ctx, actor)
-                                except Exception:
-                                    response = "Unfortunately, somthing went wrong with API"
-                                signal.alarm(0)
-                            else:
-                                api2use = key
-                                int_ctx.save_to_shared_memory(ctx, actor, api2use=api2use)
-                                response = f"""I need to use {api2use} to handle your request. Do you approve?"""
-                            break
-
-            except KeyError:
-                api2use = "generative_lm"
-                int_ctx.save_to_shared_memory(ctx, actor, api2use=api2use)
+        return response
+
+
+def complete_subtask(ctx: Context, actor: Actor, *args, **kwargs) -> str:
+    if not ctx.validation:
+        shared_memory = int_ctx.get_shared_memory(ctx, actor)
+        api2use = shared_memory.get("api2use", None)
+        subtask_results = shared_memory.get("subtask_results", {})
+        step = shared_memory.get("step", 0)
+        if api2use:
+            timeout = api_conf[api2use]["timeout"]
+            signal.signal(signal.SIGALRM, timeout_handler)
+            signal.alarm(timeout)
+            try:
+                logger.info(f"api name: {api2use}")
                 response = globals()[f"{api2use}_response"](ctx, actor)
-        else:
-            response = None
+            except Exception:
+                response = "Unfortunately, something went wrong with API"
+            signal.alarm(0)
+        logger.info(f"subtask response: {response}")
+        subtask_results[str(step)] = response
+        logger.info(f"subtask result: {subtask_results}")
+        int_ctx.save_to_shared_memory(ctx, actor, subtask_results=subtask_results)
+        return response
```

Comment:
you definetely have no return in some cases but you should because otherwise it would be a none but it might cause problems in agent. see other comments also about it